File: clients/ios/Sources/SDUIModels.swift
import Foundation

/// Mirrors KarmaDots SDUI wire format (subset).
/// Also accepts legacy/stub payloads: `version` (instead of `schemaVersion`), `screen.screenId`, `screen.sections` with `{ type, text }`.
struct SDUIEnvelope: Codable {
    let schemaVersion: Int
    let minAppVersion: String?
    let screen: SDUIScreen

    enum CodingKeys: String, CodingKey {
        case schemaVersion
        case version
        case minAppVersion
        case screen
    }

    init(schemaVersion: Int, minAppVersion: String?, screen: SDUIScreen) {
        self.schemaVersion = schemaVersion
        self.minAppVersion = minAppVersion
        self.screen = screen
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        if let sv = try c.decodeIfPresent(Int.self, forKey: .schemaVersion) {
            schemaVersion = sv
        } else if let v = try c.decodeIfPresent(Int.self, forKey: .version) {
            schemaVersion = v
        } else {
            throw DecodingError.keyNotFound(
                CodingKeys.schemaVersion,
                .init(codingPath: c.codingPath, debugDescription: "Missing schemaVersion or version")
            )
        }
        minAppVersion = try c.decodeIfPresent(String.self, forKey: .minAppVersion)
        screen = try c.decode(SDUIScreen.self, forKey: .screen)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(schemaVersion, forKey: .schemaVersion)
        try c.encodeIfPresent(minAppVersion, forKey: .minAppVersion)
        try c.encode(screen, forKey: .screen)
    }
}

struct SDUIScreen: Codable {
    let id: String
    let title: String?
    let components: [UIComponent]?

    enum CodingKeys: String, CodingKey {
        case id
        case screenId
        case title
        case components
        case sections
    }

    init(id: String, title: String?, components: [UIComponent]?) {
        self.id = id
        self.title = title
        self.components = components
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        if let i = try c.decodeIfPresent(String.self, forKey: .id) {
            id = i
        } else if let sid = try c.decodeIfPresent(String.self, forKey: .screenId) {
            id = sid
        } else {
            throw DecodingError.keyNotFound(
                CodingKeys.id,
                .init(codingPath: c.codingPath, debugDescription: "Missing id or screenId")
            )
        }
        title = try c.decodeIfPresent(String.self, forKey: .title)
        if let comps = try c.decodeIfPresent([UIComponent].self, forKey: .components) {
            components = comps
        } else if let secs = try c.decodeIfPresent([SDUIStubSection].self, forKey: .sections) {
            components = secs.map(\.asUIComponent)
        } else {
            components = nil
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encodeIfPresent(title, forKey: .title)
        try c.encodeIfPresent(components, forKey: .components)
    }
}

/// Wire shape used by some deployed stubs: `{ "type": "text", "text": "..." }`.
private struct SDUIStubSection: Decodable {
    let type: String
    let text: String?
    let content: String?

    var asUIComponent: UIComponent {
        UIComponent(
            type: type,
            layout: nil,
            spacing: nil,
            style: nil,
            content: content ?? text,
            label: nil,
            icon: nil,
            textStyle: nil,
            children: nil,
            action: nil
        )
    }
}

struct UIComponent: Codable, Identifiable {
    var id: String { "\(type)|\(label ?? "")|\(content ?? "")|\(icon ?? "")" }
    let type: String
    let layout: String?
    let spacing: Double?
    let style: ComponentStyle?
    let content: String?
    let label: String?
    let icon: String?
    let textStyle: TextStyle?
    let children: [UIComponent]?
    let action: SDUIAction?
}

struct TextStyle: Codable {
    let fontSize: Double?
    let fontWeight: String?
    let textAlign: String?
    let color: String?
}

struct ComponentStyle: Codable {
    let padding: PaddingSpec?
    let margin: MarginSpec?
    let height: DimensionSpec?
    let width: DimensionSpec?
    /// Full-width bar fill (e.g. volunteer home mockup: #FFCA10, #7A95CA).
    let backgroundColor: String?
    /// Pill / button corner radius (dp).
    let borderRadius: Double?
}

struct PaddingSpec: Codable {
    let all: Double?
    let top: Double?
    let bottom: Double?
    let left: Double?
    let right: Double?
}

struct MarginSpec: Codable {
    let top: Double?
    let bottom: Double?
    let left: Double?
    let right: Double?
    let all: Double?
}

struct DimensionSpec: Codable {
    let value: Double?
    let unit: String?
}

struct SDUIAction: Codable {
    let type: String
    let target: String?
    /// Server sends string-keyed payloads (retreatId, date, …).
    let payload: [String: String]?
}

extension UIComponent {
    func stringPayload(_ key: String) -> String? {
        action?.payload?[key]
    }
}

File: clients/android/jewelheart-admin/src/main/java/org/jewelheart/admin/SduiModels.kt
package org.jewelheart.admin

import com.google.gson.annotations.SerializedName

data class SduiEnvelope(
    @SerializedName("schemaVersion") val schemaVersion: Int,
    @SerializedName("minAppVersion") val minAppVersion: String?,
    @SerializedName("screen") val screen: SduiScreen,
)

data class SduiScreen(
    @SerializedName("id") val id: String,
    @SerializedName("title") val title: String?,
    @SerializedName("components") val components: List<UiComponent>?,
    @SerializedName("metadata") val metadata: SduiScreenMetadata?,
)

data class SduiScreenMetadata(
    @SerializedName("stickyHeader") val stickyHeader: Boolean?,
    @SerializedName("stickyHeaderComponents") val stickyHeaderComponents: List<UiComponent>?,
    @SerializedName("stickyFooter") val stickyFooter: Boolean?,
    @SerializedName("stickyFooterComponents") val stickyFooterComponents: List<UiComponent>?,
    @SerializedName("homeSplitLayout") val homeSplitLayout: Boolean?,
    @SerializedName("layoutFlat") val layoutFlat: Boolean?,
    @SerializedName("buildStamp") val buildStamp: String?,
    @SerializedName("filterState") val filterState: SduiFilterState?,
)

data class SduiFilterState(
    @SerializedName("daysAll") val daysAll: String?,
    @SerializedName("selectedDays") val selectedDays: String?,
    @SerializedName("daysPrev") val daysPrev: String?,
    @SerializedName("jobsAll") val jobsAll: String?,
    @SerializedName("selectedJobs") val selectedJobs: String?,
    @SerializedName("jobsPrev") val jobsPrev: String?,
)

data class UiComponent(
    @SerializedName("type") val type: String,
    @SerializedName("layout") val layout: String?,
    @SerializedName("spacing") val spacing: Double?,
    @SerializedName("style") val style: ComponentStyle?,
    @SerializedName("content") val content: String?,
    @SerializedName("label") val label: String?,
    @SerializedName("icon") val icon: String?,
    @SerializedName("textStyle") val textStyle: TextStyle?,
    @SerializedName("children") val children: List<UiComponent>?,
    @SerializedName("action") val action: SduiAction?,
)

data class TextStyle(
    @SerializedName("fontSize") val fontSize: Double?,
    @SerializedName("fontWeight") val fontWeight: String?,
    @SerializedName("textAlign") val textAlign: String?,
    @SerializedName("color") val color: String?,
)

data class ComponentStyle(
    @SerializedName("padding") val padding: PaddingSpec?,
    @SerializedName("margin") val margin: MarginSpec?,
    @SerializedName("height") val height: DimensionSpec?,
    @SerializedName("minHeight") val minHeight: DimensionSpec?,
    @SerializedName("width") val width: DimensionSpec?,
    @SerializedName("backgroundColor") val backgroundColor: String?,
    @SerializedName("borderColor") val borderColor: String?,
    @SerializedName("borderRadius") val borderRadius: Double?,
    @SerializedName("fullBleed") val fullBleed: Boolean?,
    @SerializedName("elevation") val elevation: Double?,
    @SerializedName("buttonVariant") val buttonVariant: String?,
    @SerializedName("equalWidthChildren") val equalWidthChildren: Boolean?,
    @SerializedName("parentCentered") val parentCentered: Boolean?,
    @SerializedName("flexGrow") val flexGrow: Boolean?,
    @SerializedName("wrapChildren") val wrapChildren: Boolean?,
    @SerializedName("multiline") val multiline: Boolean?,
    @SerializedName("navIcon") val navIcon: Boolean?,
    @SerializedName("instructionBarBleed") val instructionBarBleed: Boolean?,
    @SerializedName("maxHeight") val maxHeight: DimensionSpec?,
    @SerializedName("homeActionPill") val homeActionPill: Boolean?,
    @SerializedName("homeActionPillFullWidth") val homeActionPillFullWidth: Boolean?,
    @SerializedName("jobListFrame") val jobListFrame: Boolean?,
    @SerializedName("noWrap") val noWrap: Boolean?,
)

data class PaddingSpec(
    @SerializedName("all") val all: Double?,
    @SerializedName("top") val top: Double?,
    @SerializedName("bottom") val bottom: Double?,
    @SerializedName("left") val left: Double?,
    @SerializedName("right") val right: Double?,
)
data class MarginSpec(@SerializedName("top") val top: Double?)
data class DimensionSpec(@SerializedName("value") val value: Double?)

data class SduiAction(
    @SerializedName("type") val type: String,
    @SerializedName("target") val target: String?,
    @SerializedName("payload") val payload: Map<String, Any>?,
)

data class NavSnapshot(
    val screenId: String,
    val retreatId: String?,
    val extraParams: Map<String, String>,
)
